diff --git a/policy/modules/services/docker.fc b/policy/modules/services/docker.fc
index a5d0868e4..892b6ab93 100644
--- a/services/docker.fc
+++ b/services/docker.fc
@@ -8,3 +8,14 @@
 /usr/bin/containerd-stress	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
 /usr/sbin/dockerd	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
 /usr/sbin/docker-proxy	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
+/run/torcx/unpack/docker/bin/docker	--	gen_context(system_u:object_r:dockerc_exec_t,s0)
+/run/torcx/unpack/docker/bin/dockerd	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
+/run/torcx/unpack/docker/bin/docker-proxy	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
+/run/torcx/unpack/docker/bin/containerd	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
+/run/torcx/unpack/docker/bin/containerd-shim	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
+/run/torcx/unpack/docker/bin/containerd-shim-runc-v1	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
+/run/torcx/unpack/docker/bin/containerd-shim-runc-v2	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
+/run/torcx/unpack/docker/bin/containerd-stress	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
+/run/torcx/unpack/docker/sbin/dockerd	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
+/run/torcx/unpack/docker/sbin/docker-proxy	--	gen_context(system_u:object_r:dockerd_exec_t,s0)
+/run/torcx/unpack/docker/bin/tini --	gen_context(system_u:object_r:dockerd_exec_t,s0)
diff --git a/policy/modules/services/docker.te b/policy/modules/services/docker.te
index ffaacff0b..b42670a33 100644
--- a/services/docker.te
+++ b/services/docker.te
@@ -168,3 +168,13 @@ userdom_search_user_runtime(dockerc_user_t)
 xdg_search_data_dirs(dockerc_user_t)
 
 container_stream_connect_user_containers(dockerc_user_t)
+
+optional_policy(`
+	gen_require(`
+		type proc_kcore_t;
+	')
+
+	allow dockerd_t proc_kcore_t:file mounton;
+')
+
+allow dockerd_t initrc_t:unix_stream_socket { accept read write };
-- 
2.35.1

