diff --git a/configure.ac b/configure.ac
index 6b03a34..ea0bfc0 100644
--- a/configure.ac
+++ b/configure.ac
@@ -430,13 +430,10 @@ selinux_save_CPPFLAGS="${CPPFLAGS}"
 CFLAGS="${CFLAGS} -I${sepol_devel_incdir} -I${selinux_devel_incdir}"
 CPPFLAGS="${CPPFLAGS} -I${sepol_devel_incdir} -I${selinux_devel_incdir}"
 AC_CHECK_HEADER([sepol/sepol.h], , AC_MSG_ERROR([could not find sepol headers at $sepol_devel_incdir - make sure libsepol-devel is installed]))
-AC_CHECK_LIB([sepol], [sepol_policydb_read], ,
-             AC_MSG_ERROR([could not find libsepol at $sepol_devel_libdir]))
+LIBS="-lsepol $LIBS"
 AC_CHECK_HEADER([selinux/selinux.h], , AC_MSG_ERROR([could not find selinux headers at $selinux_devel_incdir - make sure libselinux-devel is installed]))
 AC_CHECK_HEADER([selinux/context.h], , AC_MSG_ERROR([could not find selinux headers at $selinux_devel_incdir - make sure libselinux-devel is installed]))
-AC_CHECK_LIB([selinux], [selinux_policy_root], ,
-AC_MSG_ERROR([could not find libselinux at $selinux_devel_libdir]),
-             -lsepol)
+LIBS="-lselinux $LIBS"
 SELINUX_LIB_FLAG="-L${sepol_devel_libdir} -L${selinux_devel_libdir}"
 CFLAGS="${selinux_save_CFLAGS}"
 CPPFLAGS="${selinux_save_CPPFLAGS}"
@@ -448,8 +445,6 @@ AC_ARG_ENABLE(sepol-src,
               sepol_srcdir="")
 if test "x${sepol_srcdir}" = "x"; then
    sepol_srcdir=${sepol_devel_libdir}
-   AC_CHECK_FILE([${sepol_srcdir}/libsepol.a],,
-      AC_MSG_ERROR([make sure libsepol-static is installed]))
 else
    AC_MSG_CHECKING([for compatible sepol source tree])
    sepol_version=${sepol_srcdir}/VERSION

